models/components.go

```diff
@@ -24,7 +28,93 @@ type Component struct {
 	References    *GeneralReferences      `yaml:"references" json:"references"`
 	Verifications *VerificationReferences `yaml:"verifications" json:"verifications"`
 	Satisfies     *SatisfiesList          `yaml:"satisfies" json:"satisfies"`
-	SchemaVersion float32                 `yaml:"schema_version" json:"schema_version"`
+	SchemaVersion semver.Version          `yaml:"-" json:"-"`
+}
+
+type componentLoadError struct {
+	message string
+}
+
+// Error implements the error interface by simply returning the message as a string.
+func (e componentLoadError) Error() string {
+	return e.message
+}
+
+// UnmarshalYAML is a overridden implementation of YAML parsing the component.yaml
+// This method is similar to the one found here: http://choly.ca/post/go-json-marshalling/
+// This is necessary because we want to have backwards compatibility with parsing the old types of version 2.0
+// (type =float).
+// To compensate for that, we have to hand roll our own UnmarshalYAML that can decide what to do for parsing
+// the older version of type float and converting it into semver. In addition, we will use this logic to parse strings
+// into semver.
+func (c *Component) UnmarshalYAML(unmarshal func(v interface{}) error) error {
+	// When we call "unmarshal" callback on an object, it will call that object's "UnmarshalYAML" if defined.
+	// Since we are currently in the implementation of Component's "UnmarshalYAML", when finally we call
+	// unmarshal again, if it's on type Component, we would end up in a recursive infinite loop.
+	// To prevent this, we create a separate type, called Alias.
+	type Alias Component
+	// Create an anonymous struct with an interface{} type for the schema_version that we want to parse
+	aux := &struct {
+		SchemaVersion interface{} `yaml:"schema_version" json:"schema_version"`
+		Alias         `yaml:",inline"`
+	}{
+		Alias: (Alias)(*c),
+	}
+
+	// Call unmarshal on the new Alias type. Don't return the error yet because we want to gather more information
+	// if we can below.
+	err := unmarshal(&aux)
+
+	// Create a placeholder variable for the converted semver.
+	var ver semver.Version
+	// Create a placeholder variable for the error.
+	var versionErr error
+
+	// Store the version value for conciseness.
+	value := aux.SchemaVersion
+
+	// Try to cast the value from interface{} to certain types.
+	switch v := value.(type) {
+	// For float types, which are the old types, we need to upcast it to semver if it's an older version.
+	case float32, float64:
+		switch v {
+		// Schema Version started being documented with "2.0".
+		// We should be able to parse it for backwards compatibility.
+		// All future versioning should be in semver format already.
+		case 2.0:
+			ver = semver.MustParse("2.0.0")
+		// If not the older version, it needs to be in semver format, send an error.
+		default:
+			return componentLoadError{fmt.Sprintf(constants.ErrVersionNotInSemverFormatf, v)}
+
+		}
```

Comment:
I moved it only allowing float for version `2.0`. To stop a user from continuing to use floats, if they pass `3.0` it will error out instead.